### PR TITLE
add signup, verify account, login, refresh token

### PR DESCRIPTION
**Pull request by ledat139**
# ✅ Checklist Tự review
## 🧹 Code sạch & format chuẩn
- [ ] Code sạch, dễ đọc, theo convention của team (naming, cấu trúc)
- [ ] Đã chạy `go fmt`, `goimports`, `golangci-lint` (nếu có)
- [ ] Không còn đoạn code tạm, `fmt.Println`, `log.Println`, comment debug
- [ ] Không còn biến, hàm, import không sử dụng
- [ ] Hàm không quá dài, có thể tách nhỏ nếu vượt quá 50 dòng
- [ ] Code không lặp lại, đã tái sử dụng hợp lý

## 🧠 Logic & Xử lý lỗi
- [ ] Logic đúng yêu cầu, không dư thừa, rõ ràng
- [ ] Xử lý đầy đủ các trường hợp `nil`, `error`, `invalid input`
- [ ] Mỗi thao tác có thể lỗi (I/O, DB, parse...) đều được check `err`
- [ ] Không để `err` bị silent — luôn log hoặc trả về (tùy ngữ cảnh)
- [ ] Sử dụng `context` đúng cách (nếu dùng), tránh leak
- [ ] Nếu là `goroutine`, có quản lý lifecycle và tránh leak

## 🌐 HTTP Handler (Gin)
- [ ] Sử dụng đúng HTTP method (GET, POST, PUT, DELETE…)
- [ ] Validate request (param, query, body) đầy đủ, rõ ràng
- [ ] Sử dụng binding (`ShouldBindJSON`, `ShouldBindUri`, v.v.) đúng cách
- [ ] Xử lý trả về `status code` phù hợp (`200`, `201`, `400`, `404`, `500`…)
- [ ] Response rõ ràng, có `message`, `data`, `error` (nếu cần)
- [ ] Không leak thông tin nhạy cảm trong response lỗi

## 🗃️ Struct & Layer
- [ ] Struct có annotation JSON đầy đủ (`json:"field_name"`)
- [ ] Sử dụng DTO/input/output rõ ràng, không lẫn lộn model DB và response
- [ ] Đã tách rõ các layer: handler → service → repository (nếu theo clean code)
- [ ] Middleware viết rõ ràng, dễ đọc, không làm chậm hệ thống

## 🧪 Unit Test & Mock
- [ ] Có test cho handler/service chính (ít nhất happy case và edge case)
- [ ] Nếu dùng DB/service ngoài, có mock hoặc test isolation
- [ ] Có test validate đầu vào (request, param, query)
- [ ] Không dùng test hardcode giá trị không ổn định (vd: time.Now() mà không mock)
- [ ] Tên test rõ ràng: `TestFunction_Scenario_Expected`

## 🔐 Security & Config
- [ ] Không hardcode secret, token, config trong code
- [ ] Không để lộ thông tin nhạy cảm qua log hoặc panic
- [ ] Validate và sanitize đầu vào người dùng
- [ ] Nếu có auth, kiểm tra middleware hoạt động đúng

## 🤖 Sử dụng GitHub Copilot
- [ ] Đã kiểm tra kỹ mọi đoạn code được gợi ý từ GitHub Copilot
- [ ] Không để lại phần comment hoặc TODO do Copilot sinh ra
- [ ] Không commit đoạn code mà chưa hiểu rõ logic hoặc mục đích
- [ ] Đảm bảo mọi function/code block từ Copilot phù hợp với context dự án
- [ ] Tối ưu lại code được gợi ý nếu cần — tránh lạm dụng hoặc copy nguyên mẫu

## Related Tickets
- ticket redmine link

## WHAT
- Xây dựng 3 service: Auth Service, Mail Service, User Service.
- Auth Service: xử lý đăng ký, đăng nhập, refresh token, xác thực tài khoản.
- Mail Service: nhận event từ Kafka và gửi email xác thực.
- User Service:  tạo mới người dùng khi đăng ký.

## HOW
- Auth Service:
  - Thêm logic đăng ký (SignUp):
    - Kiểm tra email đã tồn tại.
    - Hash mật khẩu bằng bcrypt.
    - Gọi User Service để tạo hồ sơ người dùng.
    - Sinh JWT token xác thực và publish event qua Kafka.
  - Thêm logic đăng nhập (Authenticate):
    - Kiểm tra email, mật khẩu.
    - Sinh access token và refresh token.
  - Thêm logic (Refresh Token):
    - Xác thực refresh token.
    - Kiểm tra trạng thái kích hoạt và xác thực của người dùng.
    - Sinh access token mới.
  - Thêm logic (VerifyAccount):
    - Giải mã token từ email.
    - Cập nhật trạng thái đã xác thực.
- Mail Service:
  - Tạo Kafka consumer lắng nghe event xác minh tài khoản từ Auth Service.
  - Gửi email xác thực tài khoản sử dụng template HTML và SMTP.
- User Service:
  - Cung cấp API tạo người dùng (CreateUser).
  - Lưu thông tin tên, email, role vào database.
  - Kiểm tra trùng lặp và validate dữ liệu.
  - Tích hợp với Auth Service để tự động tạo hồ sơ khi người dùng đăng ký.

## WHY
- Trước đây hệ thống chưa có luồng xác thực, quản lý hồ sơ người dùng và gửi email xác minh.
- Việc tách thành 3 service giúp hệ thống dễ mở rộng, bảo trì và có thể triển khai độc lập.
- Bổ sung kiểm tra trạng thái và xác thực giúp tăng tính bảo mật, tránh truy cập trái phép.
- Tự động gửi email xác minh giúp cải thiện trải nghiệm và đảm bảo email người dùng hợp lệ.

## Notes
- Kiến thức tìm hiểu thêm:
  - Microservice
  - Sử dụng Kafka làm message broker giữa các service.
  - Tích hợp gửi email qua SMTP với template HTML.
  - Unit test cho AuthUseCase để đảm bảo logic hoạt động đúng.

### CODE REVIEW

**Review by truongtamthanh2004 — APPROVED** [review]

File: services/auth-service/internal/handler/auth_handler.go [hunk 1]
```diff
@@ -1 +1,145 @@
...
+	uc usecase.AuthUsecase
+}
+
+func NewAuthHandler(uc usecase.AuthUsecase) *AuthHandler {
+	return &AuthHandler{uc: uc}
+}
+
+func (h *AuthHandler) SignUp(c *gin.Context) {
+	var req dto.SignupRequest
+	if err := c.ShouldBindJSON(&req); err != nil {
```
Comment (chienpv-3590) [inline services/auth-service/internal/handler/auth_handler.go:24]: Nên bổ sung kiểm tra dữ liệu đầu vào chi tiết hơn (email format, password strength, v.v.) thay vì chỉ dùng ShouldBindJSON.

File: services/auth-service/internal/handler/auth_handler.go [hunk 2]
```diff
@@ -1 +1,145 @@
...
+
+	err := h.uc.VerifyAccount(c.Request.Context(), tokenString)
+	if err != nil {
+		switch err.Error() {
+		case constant.ErrGetUserFailed:
+			c.JSON(http.StatusNotFound, gin.H{"message": err.Error()})
+		case constant.ErrInvalidToken, constant.ErrUserAlreadyVerified:
+			c.JSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		case constant.ErrFailedToUpdateUser:
+			c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
```
Comment (chienpv-3590) [inline services/auth-service/internal/handler/auth_handler.go:74]: Tránh trả về lỗi quá chi tiết cho client với các lỗi hệ thống, chỉ nên trả về thông báo chung.

File: services/auth-service/internal/kafka/producer.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	payload, err := json.Marshal(event)
+	if err != nil {
+		return errors.New(constant.ErrMarshalRequest)
+	}
+
+	if err := p.writer.WriteMessages(ctx, kafka.Message{
+		Key:   []byte(email),
+		Value: payload,
+	}); err != nil {
+		fmt.Println("Failed to write message:", err)
```
Comment (chienpv-3590) [inline services/auth-service/internal/kafka/producer.go:49]: Thay vì dùng fmt.Println, nên dùng thư viện logging như log hoặc các logger chuyên nghiệp (zap, zerolog…) để kiểm soát mức độ log và lưu trữ log tốt hơn.

File: services/auth-service/internal/handler/auth_handler.go (hunk 1 above)
Comment (ledat139) [reply services/auth-service/internal/handler/auth_handler.go:24]: format email em đã thêm bằng tag binding:email ở dto rồi ạ

**Comment by ledat139** [comment]
READY
